src/main/java/edu/harvard/iq/dataverse/export/InternalExportDataProvider.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package edu.harvard.iq.dataverse.export;
+
+import java.io.InputStream;
+import java.util.Optional;
+
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonArrayBuilder;
+import javax.json.JsonObject;
+import javax.json.JsonObjectBuilder;
+
+import edu.harvard.iq.dataverse.DOIDataCiteRegisterService;
+import edu.harvard.iq.dataverse.DataCitation;
+import edu.harvard.iq.dataverse.DataFile;
+import edu.harvard.iq.dataverse.DatasetVersion;
+import edu.harvard.iq.dataverse.FileMetadata;
+import io.gdcc.spi.export.ExportDataProvider;
+import edu.harvard.iq.dataverse.util.bagit.OREMap;
+import edu.harvard.iq.dataverse.util.json.JsonPrinter;
+import edu.harvard.iq.dataverse.util.json.JsonUtil;
+
+/**
+ * Provides all data necessary to create an export
+ * 
+ */
+public class InternalExportDataProvider implements ExportDataProvider {
```

Comment:
Is this interface and implementation really needed? Why are there `internal` export Formats and why are those handled different than the others? IMHO: The same result can be achieved without this interface/implementation. Sure this approach protects the internal `DataSet` object from access by an Exporter implementation, but is this needed?  IMHO, this just makes development of new exports more complex to implement (or forces me to use reflection to get the objects ;))

Reply:
Yes this is needed. Let me explain: the internal model is using lots of database annotations. If we directly move these classes into the `dataverse-spi` package, that means everyone using it will inherit the dependencies. This is not how you design an API like this: you want to decouple internal and external code.

Usually one would use data transfer objects for this that are shipped in the API package and where the main code has converters in place to move the data in and out. However, this is a larger task then one might anticipate and could not be done on current funding.

So the technical compromise is to use JSON / JSON-LD / XML as data transfer, although it needs re-processing on the plugin side. This needs to be shipped somehow, and the external plugin needs a clear interface definition how to access the data and what kind of data is expected.

On the other hand, the main code must provide an actual implementation of the interface, otherwise the external plugin will never be able to query for the data. So this is the implementation of this interface, which is internal to the main code - it will never be shown to the external plugin, which is always relying on the interface.

It's not great, it's not perfect, but very reliable. Even if you would go for DTOs, this provider thing might still be there, as it could keep hold of the conversion process from the DB model to the external model.

Hope this makes all more sense now.